pyart/core/grid.py

```diff
@@ -81,18 +284,57 @@ def add_field(self, field_name, field_dict, replace_existing=False):
             raised if field_name already exists.
 
         """
-
-        # grid fields should have dimensions of (nz, ny, nx)
-        nz, ny, nx = self.fields[self.fields.keys()[0]]['data'].shape
-
         # checks to make sure input field dictionary is valid
         if 'data' not in field_dict:
             raise KeyError('Field dictionary must contain a "data" key')
         if field_name in self.fields and replace_existing is False:
             raise ValueError('A field named %s already exists' % (field_name))
-        if field_dict['data'].shape != (nz, ny, nx):
+        if field_dict['data'].shape != (self.nz, self.ny, self.nx):
             raise ValueError('Field has invalid shape')
 
         self.fields[field_name] = field_dict
 
-        return
+
+def _point_data_factory(grid, coordinate):
+    """ Return a function which returns the locations of all points.  """
+    def _point_data():
+        """ The function which returns the locations of all points. """
+        reg_x = grid.regular_x['data']
+        reg_y = grid.regular_y['data']
+        reg_z = grid.regular_z['data']
+        if coordinate == 'x':
+            return np.tile(reg_x, (len(reg_z), len(reg_y), 1)).swapaxes(2, 2)
+        elif coordinate == 'y':
+            return np.tile(reg_y, (len(reg_z), len(reg_x), 1)).swapaxes(1, 2)
+        else:
+            assert coordinate == 'z'
+            return np.tile(reg_z, (len(reg_x), len(reg_y), 1)).swapaxes(0, 2)
+    return _point_data
+
+
+def _point_lon_lat_data_factory(grid, coordinate):
+    """ Return a function which returns the geographic locations of points. """
+    def _point_lon_lat_data():
+        """ The function which returns the geographic point locations. """
+        x = grid.point_x['data']
+        y = grid.point_y['data']
+        projparams = grid.projection.copy()
+        if projparams.pop('_include_lon_0_lat_0', False):
+            projparams['lon_0'] = grid.origin_longitude['data'][0]
+            projparams['lat_0'] = grid.origin_latitude['data'][0]
+        geographic_coords = cartesian_to_geographic(x, y, projparams)
+        # set the other geographic coordinate
+        if coordinate == 0:
+            grid.point_latitude['data'] = geographic_coords[1]
```

Comment:
Yes, [1] is correct in this context, but it is a bit confusing.  What is being done here is when the point_latitude or point_longitude 'data' key is evaluated 'data' for the other attribute is also calculated and stored.  This prevents the inverse map projection from having the be calculated twice.  Also, this prevents the point_longitude and point_latitude attributes from containing data from two differing projections.  Either both point_latitude['data'] and point_longitude['data'] are un-evaluated or both have been evaluated using the same map projection.

I'll expand the comment before this line to make this more clear